### PR TITLE
DEPR: Remove NumericIndex

### LINKED ISSUE #42717: DEPR: Remove Int64Index, UInt64Index, Float64Index and Index._is_backward_compat_public_numeric_index
In Pandas 2.0 NumericIndex should be the default numerical index and `Int64Index`, `UInt64Index` and `Float64Index` should be removed. See #41153 & 41272 for more discussion.

#41153 adds an internal attribute `_is_backward_compat_public_numeric_index` to the `Index` class (and its numeric child classes). This is only used to keep track of whether an index is a `NumericIndex` or one of its child indexes (`Int64Index`, `UInt64Index` and `Float64Index`) in order to maintain backwards compatability in Pandas 1.x. It will not be needed in pandas 2.0 after the child classes are removed and should removed in pandas 2.0 also.

### LINKED ISSUE #51020: Include NumericIndex at top level and document it for version 2.0
With the removal of `IntegerIndex`, `FloatIndex`, etc., and now promoting its replacement `NumericIndex`, there are a couple of things to consider:

1. I think that `NumericIndex` should appear in `pandas.__init__.py` like the other Index types so that `pd.NumericIndex` works.
2. We should document it, since we refer to it in the documentation.  It appears as a result in examples, and we refer to it, e.g. here: 

https://pandas.pydata.org/docs/dev/reference/api/pandas.Index.html?highlight=numericindex

@topper-123 may be tracking this based on comments in #50908

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes #42717
- [x] closes #51020

This finishes the removal of `NumericIndex` from the code base. Note that this PR builds atop of #51132, so you might want to merge that first.

Everything should be finished now after this PR, e.g. the doc updates  have been merged (#51111) and the doc string has been updated (#51089). I'll have a final look once this is merged, but this should pretty much finish this issue.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -5375,6 +5362,7 @@ def identical(self, other) -> bool:
                 for c in self._comparables
             )
             and type(self) == type(other)
+            and self.dtype == other.dtype
```
Comment (topper-123) [inline pandas/core/indexes/base.py:5365]: Previously the index type could differentiate e.g. `Index([1, 2, 3], dtype=object)` and `Index([1, 2, 3], dtype="int64")` (the latter actually being a `NumericIndex`), but we now need to be more precise.

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -539,7 +538,6 @@ def __new__(
 
         klass = cls._dtype_to_subclass(arr.dtype)
 
-        # _ensure_array _may_ be unnecessary once NumericIndex etc are gone
         arr = klass._ensure_array(arr, arr.dtype, copy=False)
```
Comment (phofl) [inline pandas/core/indexes/base.py:541]: It's not unnecessary?

**Review by phofl — APPROVED** [review]
small comment (can also be addressed in a follow up), otherwise lgtm

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/indexes/base.py:541]: I had to remove “NumericIndex” and had to decide what to do with comment and opted to delete it. IMO _ensure_array could be refactored, but so could a lot of other things in pandas and _ensure_array is not *that* bad IMO anymore compared to other things, so doesn’t merit a comment In the code base IMO.

This was my reasoning, others could of course have made a different judgement call…So it’s not a strong opinion, and I’m fine with it being refactored, but also it not being refactored :-)

Comment (phofl) [reply pandas/core/indexes/base.py:541]: Yeah no worries, just wasn't sure if this was still being True or not

**Comment by mwtoews** [comment]
This removal is the cause of a regression unpickling files from "pandas<2", #53300

**Comment by Aaron1993** [comment]
This remove action caused a bug `ModuleNotFoundError: No module named 'pandas.core.indexes.numeric` when use `pandas>=2.0.0 `

**Pull request merged**